app/models/subscription.rb

```diff
@@ -15,8 +15,9 @@ def search_criteria_to_h
   end
 
   def set_reference
-    self.reference ||= loop do
-      reference ||= SecureRandom.hex(8)
+    return if reference.present?
+    self.reference = loop do
```

Comment:
I can limit the loop attempts, but I'm not sure I see the value in that. It's also extremely unlikely that the same token will be generated more than once, so I don't see any value in extending the character count. 

Why do you think this could loop indefinitely? The only way that would happen is 
- if the token is empty, but we make sure we only loop if it's not
- if SecureRandom keeps generating the same number!

https://github.com/dxw/teacher-vacancy-service/blob/feature/job-seeker-alerts/spec/models/subscription_spec.rb#L72

Reply:
I was looking for a unit test that explicitly describes `set_reference` and covers the unhappy path when there is a preexisting duplicate record. What we do have is a test that skips the method itself and does check the multiple steps of creation journey work together. Which is good to have but maybe not instead of. On trying to write the test I think we need, it's not easy to simulate which is a smell to me.

```
describe '#set_reference' do 
  context 'when a record already exists with the random reference and email address'
    before do
      create(:subscription, id: '123', email: 'foo@email.com')
      allow(SecureRandom).to receive(:hex).and_return('123')
    end
    it 'tries again until it creates one'
      # quite hard to test with multiple SecureRandom stubs
```

I think it is true to say that right now as the code is, it _would_ be improbable to get a duplicate at all. 

My concern is that while we are _currently_ confident to make that statement, we cannot ensure that it will be the case forever. A new feature or change could come along at any point and we'd be reliant on that developer to come and check this untested scenario to make sure it continues to work without unintended consequences. Ultimately the worst that could happen is we accidentally make a change that never breaks out of the loop and the user has to wait until their web request times out.

In an effort to keep it simple, how about a compromise of replace the loop and guard for always setting a 32 character hex reference (instead of 8). We could then use a presenter/helper to only show the first 8 characters to users. The chances of a duplicate should be then be even lower removing the need for guard at all. Activerecord validation could then ensure uniqueness and be responsible for raising an error up to the user advising them to try again.

Reply:
I've just picked this up and could we not do a SHA-256 hash of a combination of the user's email address and the current datetime, then restrict to the first 8 characters?

Something like this should ensure uniqueness:

```
self.reference = Digest::SHA256.hexdigest("#{email}-#{Time.now.to_i}")[0..7]
```